### ISSUE TITLE
Validate number: use only isFinite?

### ISSUE THREAD

**Issue by arnaudjuracek** [Chalarangelo/30-seconds-of-code#41]
> ### Validate number
> Use !isNaN in combination with parseFloat() to check if the argument is a number. Use isFinite() to check if the number is finite.
```js
const validateNumber = n => !isNaN(parseFloat(n)) && isFinite(n);
// validateNumber('10') -> true
```

---
According to the [MDN documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/isFinite), `isFinite()` returns "**`false`** if the argument is positive or negative `Infinity` or `NaN`; otherwise, **`true`**". 

```js
isFinite(Infinity);  // false
isFinite(NaN);       // false
isFinite(-Infinity); // false

isFinite(0);         // true
isFinite(2e64);      // true
isFinite(910);       // true

isFinite(null);      // true, would've been false with the 
                     // more robust Number.isFinite(null)

isFinite('0');       // true, would've been false with the 
                     // more robust Number.isFinite("0")
```

So technically, the `!isNan()` isn't required and you could shorten your `validateNumber`:
```js
const validateNumber = n => isFinite(n);
// validateNumber('10') -> true
```

**Comment by Chalarangelo** [comment]
I think I based the snippet on [this SO answer](https://stackoverflow.com/a/1830844/1650200). It might be outdated or there might be something I don't understand. Does anyone know of any cases where `Number.isFinite()` alone would fail?

**Comment by fesebuv** [comment]
🤔  what about? 
```
const validateNumber = n => typeof n === 'number';
```

**Comment by atomiks** [comment]
@fesebuv I believe the function is meant to allow a number string as well.

Also OP, what about booleans?

```javascript
const validateNumber = n => isFinite(n)
validateNumber(true) // true
```

**Comment by elderhsouza** [comment]
@atomiks @fesebuv @Chalarangelo 

I think this would do:
```js
const validateNumber = n => isFinite(parseFloat(n))
```

**Comment by Chalarangelo** [comment]
Yes, the function is meant to allow for strings containing numerical data, such as `'10'` for example. @elderhsouza I think this should probably work as intended. Can anyone find any test cases that break this? Otherwise, we should get a PR going and change the current solution to this.

**Comment by atomiks** [comment]
@elderhsouza 

```javascript
validateNumber('20px') // true
```

**Comment by Chalarangelo** [comment]
@atomiks technically speaking `'20px'` is a numeric value in CSS. 😛

**Comment by elderhsouza** [comment]
@Chalarangelo @atomiks 

I'm using this to test the cases, and to be fair I now think @Chalarangelo was right all along, the first option doesn't break with any test cases so far:

```js
isNumber = n => !isNaN(parseFloat(n)) && isFinite(n)

test = [Infinity, NaN, -Infinity, 0, 2e64, 910, null, '0', undefined, true, '20px', 50, '0xFFFFFF', 0xFFFFFF, '0xfff']
results = test.map(isNumber)

console.table([test, results])
```

**Comment by Chalarangelo** [comment]
I told you guys where I think I found the original, but the clarification why it works is unclear on SO. Can anyone figure it out and explain it better so we can update the snippet's description?

**Comment by Ephys** [comment]
`isNumber([1])` returns true.

I think the only safe way to do it is to explicitly whitelist number & strings

```javascript
isNumber = n => {
  const type = typeof n;
  if (type !== 'string' && type !== 'number') {
    return false;
  }

  return isFinite(n);
}
```

**Comment by Chalarangelo** [comment]
@Ephys `isNumber([1,2])` returns `false`, though. I mean an array with just one number sounds reasonably valid to me, as `parseFloat([1])` will work properly. JS is weird when it comes to stuff like this.

**Comment by Ephys** [comment]
@Chalarangelo That's due to coercion (the array is stringified into `"1"` then parsed as an int -- notice how `parseFloat([1,2])` works too for the same reason) and that has caused more bugs than I can ever count. But if you want to consider arrays and above to be numbers it's up to you and likely depends on what you're trying to achieve :)

**Comment by Chalarangelo** [comment]
@Ephys you make a pretty good point. This can cause trouble, but limiting to specific data types can be troublesome as well. Generally the parsing system of JS is terrible, I mean `parseFloat('123I am a little lamb')` returns `123` for no apparent reason. I know *why* it works and I know this is the defined behavior, but I hate it and there are so many edge-cases that this whole thing becomes a nightmare to deal with. 

In general, I agree, numbers and strings that can be converted to numbers *should* be considered valid by the implemented function (unless the are equal to `Infinity`), except that some strings that evaluate to numbers are not really numbers and we need to deal with those, I suppose.

**Comment by fesebuv** [comment]
@atomiks  @elderhsouza 
in that case:
```
const validateNumber = n => parseInt(n, 10) =>  typeof n === 'number';
```

**Comment by Ephys** [comment]
@Chalarangelo I definitely agree, parsing numbers in javascript is a pain with so many corner cases I lost count (which is why I restrict to usually restrict to strings/numbers and use `Number` to parse instead of parseFloat).

By the way the solution works for most cases is that you're using two different kind of number conversions that happen to exclude the wrong results of the other. If it helps I made a document comparing the possible methods to try and figure out the mess: https://gist.github.com/Ephys/4d9731d5c7a945dcc752d279eaeaa08f

**Comment by Chalarangelo** [comment]
@Ephys that's a lot of effort you put in, but unfortunately I am too tired to read through the whole thing right now. First thing in the morning, though. I really want to figure out what to do with this specific snippet as I expect it to be used by a lot of people.

@fesebuv this still returns `true` for `'20px'`, so it's not right yet.

**Comment by skatcat31** [comment]
I'm confused why this code would even use `!isNaN(x)` since `NaN == false`. That parsing isn't needed. However to check if something is a number you need to do some coercion checking to make sure. If it's coerced form is equal to it's numerical form then it is a number. Really what we want are more test cases that would be considered valid or invalid:

# Tests

```js
validateNumber('10') // true
validateNumber(false) // false
validateNumber(true) // false
validateNumber(10) // true
validateNumber(0.10) // true
validateNumber(010) // true
validateNumber(0xA) // true
validateNumber(NaN) // false
validateNumber(null) // false
validateNumber(undefined) // false
validateNumber('10%') // false, % is a construct
validateNumber('10px') // false
validateNumber(infinity) // false
validateNumber(Math.pi) // true, but is such an extreme edge case...
```

test based coding is really the only way to infallible say "This method is complete to the best of my knowledge". As new tests fail that should be valid we then know the code is invalid. Fix, iterate, evolve.

# Possible Code

```js
const validateNumber = n => isFinite(n) && parseFloat(n) && Number(n) == n;
// validateNumber('10') -> true
// validateNumber('10px') -> false
// validateNumber('10%') -> false, javascript would represent a percent as a decimal
```

**Comment by Chalarangelo** [comment]
Resolved in #101 as suggested. If any problems arise, comment below to reopen the issue.
